### PR TITLE
Microservices

### PR DESCRIPTION
**Pull request by Pelmenner**
Split project into following microservices:
- controller (db communication, central server)
- messengers (sending messages and receiving updates):
  - vk
  - tg

Communication between services is done through gRPC. For that there is a proto folder with .proto declarations and generated go code. 

To avoid unnecessary code duplication, baseMessenger was extracted into a separate module in src/messenger and imported in each messenger.

Most of the old code was not modified (only moved), but some changes outside of the communication part were made to improve the code and facilitate this switch to the new architecture.

### CODE REVIEW

**Review by Mad-Proger — CHANGES_REQUESTED** [review]

File: src/controller/debug.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+WORKDIR controller
+RUN mkdir data
+RUN go build -o /transfer_bot
+
+RUN chmod +x ./docker_entrypoint_debug.sh
+
+EXPOSE 8000 40000
+
+CMD ["./docker_entrypoint_debug.sh"]
```
[the remark this answers is not in the document]
Comment (Pelmenner) [reply src/controller/debug.Dockerfile]: Deleted docker_entrypoint_debug.sh as well as debug.Dockerfile

File: src/controller/main.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	if err != nil {
+		log.Println("could not delete file", attachment.URL, err)
+	}
+	err = os.Remove(filepath.Dir(attachment.URL))
+	if err != nil {
+		log.Println("could not delete directory", filepath.Dir(attachment.URL), err)
+	}
+}
+
+func repeatedFileCleanup(db *orm.DB) { // Why is it still unused?
```
[the remark this answers is not in the document]
Comment (Pelmenner) [reply src/controller/main.go:29]: There's still a lot to refactor, but it is not related to microservices

File: src/messengers/vk/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+WORKDIR ../..
+
+COPY proto/ proto/
+COPY messengers/messenger/*.go messengers/messenger/
+COPY messengers/vk/vk/*.go messengers/vk/vk/
+COPY messengers/vk/*.go messengers/vk/
+
+WORKDIR messengers/vk
+RUN mkdir -p /usr/local/bin/
+RUN go mod tidy
```
[the remark this answers is not in the document]
Comment (Pelmenner) [reply src/messengers/vk/Dockerfile]: Removed go mod tidy.
What do you mean by _correct_ go.mod and go.sum? If something is particularly wrong, build will just fail

File: src/messengers/vk/vk/vk.go [hunk 4]
```diff
@@ -1,34 +1,30 @@
...
 type VKMessenger struct {
-	BaseMessenger
+	*messenger.BaseMessenger
 	vk       *api.VK
 	longPoll *longpoll.LongPoll
 }
 
-func NewVKMessenger(baseMessenger BaseMessenger) *VKMessenger {
+func NewVKMessenger(baseMessenger *messenger.BaseMessenger) *VKMessenger {
 	token := os.Getenv("VK_TOKEN")
```
[the remark this answers is not in the document]
Comment (Pelmenner) [reply src/messengers/vk/vk/vk.go]: Done

File: src/messengers/vk/Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (Mad-Proger) [reply src/messengers/vk/Dockerfile]: > build will just fail

And that's not what we want, I think. We should have tests that check that no build error occurs in version that is in master

**Review by Mad-Proger — CHANGES_REQUESTED** [review]

**Review by Mad-Proger — APPROVED** [review]

**Pull request merged**